FIX `cross_validate` with multimetric scoring returns the non-failed scorers results if there is a failing scorer

simonandras opened the pull request:
#### Reference Issues/PRs
Fixes #22969.

#### What does this implement/fix? Explain your changes.
In `_scorer.py` in the `_MultimetricScorer` class the `__call__` method  is updated. Now instead of raising an exception of a failing scorer it passes the exception to the `scores` to handle later.

In `_validation.py` in the `_score` function the exception is handled at the line 789, based on the suggestion of @thomasjpfan.

#### Any other comments?
Do we need a test for the one-failing scorer case for the `cross_validate` function?

simonandras committed:
Formatting

simonandras committed:
If all scorer fails, raise an error.

simonandras committed:
Update warning message.

simonandras committed:
Update raised exception

simonandras commented on sklearn/metrics/_scorer.py:
```diff
@@ -101,12 +101,23 @@ def __call__(self, estimator, *args, **kwargs):
+                all_scorer_failed = False
+
+        if all_scorer_failed:
+            raise last_exception
```
This is for passing the `test_cross_validate_failing_scorer` test in `test_validation.py`.

thomasjpfan commented on sklearn/metrics/_scorer.py:
```diff
@@ -101,12 +101,23 @@ def __call__(self, estimator, *args, **kwargs):
-            scores[name] = score
+                all_scorer_failed = False
+
+        if all_scorer_failed:
+            raise last_exception
```
Since we are already going through the dictionary in `_score` I prefer to handle the exceptions there. This way `_MultimetricScorer` will only need to set the exceptions in `scores[name]`:

```python
for name, scorer in self._scorers.items():
    try:
        ...
    except Exception as e:
        scores[name] = e

return scores
```

And then in `_score`:

```python
# check errors in `_MultimetricScorer`
if isinstance(scorer, _MultimetricScorer):
    exceptions = {name: isinstance(e, Exception) for name: e in scores.items()}
    if exceptions:
        if error_score == "raise":
            # raise error
        else:
            warnings.warn(f"Scoring failed for...", UserWarning)
            # replace exceptions with error_score
            new_scores = {name: error_score for name in scores}
            scores.update(new_scores)
```

simonandras replied:
Thank you for the review! Pushed the changes.

thomasjpfan reviewed (commented):
Thank you for the PR!

> Do we need a test for the one-failing scorer case for the cross_validate function?

Yes we need to test the new behavior. One as a non-regression test for `cross_validate` and another one for `_MultimetricScorer` itself.

Note: Changing the behavior of `_MultimetricScorer` to not raise anymore may require updating some unit tests. This is okay since `_MultimetricScorer` is a private class.

simonandras committed:
Making the changes of the rewiev. TODO: update the tests and add new one to the new parts.

simonandras commented on sklearn/model_selection/_validation.py:
```diff
@@ -781,6 +781,23 @@ def _score(estimator, X_test, y_test, scorer, error_score="raise"):
+        if exceptions:
+            if error_score == "raise":
+                for name, e in exceptions.items():
+                    raise e
```
Im not sure if this is the best way to access an element of the dict. Actually I think it doesnt matter which exception we choose here.

thomasjpfan replied:
We can use a list instead:

```python
exceptions = [(name, e) for name, e in scores.items() if isinstance(e, Exception)]
```

This way we can raise directly here:

```python
raise exceptions[0][1]
```

`new_scores` would need to be updated slightly.

thomasjpfan commented on sklearn/model_selection/_validation.py:
```diff
@@ -781,6 +781,23 @@ def _score(estimator, X_test, y_test, scorer, error_score="raise"):
+                warnings.warn(
+                    "Scoring failed. The score on this train-test partition for "
+                    f"these parameters will be set to {error_score}. Details: \n"
+                    f"{format_exc()}",
```
I think we can not use `format_exc` here since this is not in a `except` block. We would need to look at `exceptions` for the information.

simonandras replied:
How can we show here the content of `exceptions`? I think it would make sence to show every error message here, not just the first one.

simonandras replied:
This way maybe, but with formatted `e`?

```
else:  # line 772
    for name, e in exceptions:
        scores[name] = error_score
        warnings.warn(
            "Scoring failed. The score on this train-test partition for "
            f"these parameters will be set to {error_score}. Details: \n"
            f"{e}",  # TODO: format like in case of `format_exc()`
            UserWarning,
        )
```

thomasjpfan replied:
> How can we show here the content of exceptions? I think it would make sence to show every error message here, not just the first one.

Yup. I think it makes sense to show all of them.

simonandras committed:
Update  to a list.

simonandras committed:
Added formatted exception to the warnings. All the failing scorers message will show up there.

simonandras commented on sklearn/model_selection/_validation.py:
```diff
@@ -781,6 +781,27 @@ def _score(estimator, X_test, y_test, scorer, error_score="raise"):
+                    scores[name] = error_score
+                    details = "".join(
+                        format_exception(etype=type(e), value=e, tb=e.__traceback__)
+                    )
```
I chose `format_exception` here, which using this way with `join` gives the same output as `format_exc()`, but now it uses the `e` exception istead of the exception blocks exception.

simonandras committed:
Calling the function with positional arguments instead.

thomasjpfan reviewed (commented):
Thanks for the update!

This PR still requires a non-regression test for the original issue and test to test the behavior of `_MultimetricScorer` returning a dictionary with exceptions.

simonandras committed:
added test_multimetric_scorer_returning_exceptions_in_dictianary. TODO: a non-regression test for the original issue.

simonandras committed:
Update test_cross_validate_failing_scorer to cover the failing and non-failing multimetric case.

simonandras commented:
Updated the `test_cross_validate_failing_scorer` to cover the issue with the failing and non-failing multimetric case. I think its better this way than writing a new test.  @thomasjpfan what do you think about this?
Also added a new test for `_MultimetricScorer` returning a dictionary.

thomasjpfan commented on sklearn/metrics/_scorer.py:
```diff
@@ -102,11 +102,15 @@ def __call__(self, estimator, *args, **kwargs):
+                    score = scorer(estimator, *args, **kwargs)
+                scores[name] = score
+            except Exception as e:
+                scores[name] = e
```
I know I recommended this approach, but is there any Python specific issues with saving the Exception like this?

REF: https://stackoverflow.com/questions/8238360/how-to-save-traceback-sys-exc-info-values-in-a-variable

If there is an issue, we would need to try another approach.

simonandras replied:
Based on the comments to the Stackoverflow question ("Be careful when you take the exception object or the traceback object out of the exception handler, since this causes circular references and gc.collect() will fail to collect. "), I agree, we should avoid it. 

If we choose to go that way, we have to store the type of the exception, the text and the `__traceback__`. Im not sure however how can we raise a new exception using data from a previous one.

thomasjpfan replied:
I'm okay with raising an exception in `_MultimetricScorer` and a new bool parameter in `__init__`.

> In that case we should only store the formatted exception text to use it in the warning.

In the warning situation, how do you want to pass the formatted exception text back up to `_score`?

simonandras replied:
Passing it as a string. And then in the `_score` function in the `scores.items()` we could look for strings, which would indicate the exception.

```
except Exception as e:
    if raise_exp:
        raise e
    else:
        scores[name] = format_exc()
```

thomasjpfan replied:
Let's go with a string for now. I do not see a great alternative that is less complicated.

thomasjpfan commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -2272,9 +2279,15 @@ def test_cross_validate_failing_scorer(
-                    # check the test (and optionally train score) for all
-                    # scorers that should be assigned to `error_score`.
-                    assert_allclose(results[key], error_score)
+                    if "2" in key:
```
I see that we already check `_score`, but I think using `score_2` is more explicit:

```suggestion
                    if "score_2" in key:
```

thomasjpfan reviewed (commented):
Please add an entry to the change log at `doc/whats_new/v1.1.rst` with tag |Fix|. Like the other entries there, please reference this pull request with `:pr:` and credit yourself (and other contributors if applicable) with `:user:`.

simonandras committed:
Changing '2' to more explicit 'score_2' in the test_cross_validate_failing_scorer

simonandras committed:
_MultimetricScorer now can raise an exception in __call__ or not. This is controlled by the new bool parameter. The _score function is changed accordingly. The test for the _MultimetricScorer which tested the exception returning case is replaced with the test_multimetric_scorer_exception_handling test. The test_cross_validate_failing_scorer function needs no modification.

simonandras committed:
Remove not used function from the imports.

simonandras commented on sklearn/metrics/_scorer.py:
```diff
@@ -88,11 +89,17 @@ class _MultimetricScorer:
     """
 
-    def __init__(self, **scorers):
+    def __init__(self, *, raise_exc=True, **scorers):
```
In this case we have to modify every usage, where the `_MultimetricScorer` is constructed right (from `_MultimetricScorer(**scorers)` to `_MultimetricScorer(scorers=scorers)`)?

thomasjpfan replied:
Yea. It's a private API so, so it's okay to change without worrying about backward compatibility. 

Logistically, `_MultimetricScorer` does not appear in many places so the diff should not be too big.

simonandras committed:
Changing _MultimetricScorer API: now the 'scorers' parameter is a positional (dict) instead of a depacked dict using **.

simonandras committed:
Adding changelog.

simonandras committed:
small changes to changelelog

simonandras committed:
Changing the changelog again.

simonandras committed:
Merge branch 'main' into cross_validate__with_multiple_scorer_nan_issue

simonandras committed:
Changing the number in the changelogand a small change in the test_cross_validate_failing_scorer: checking for '_score_2' instead of 'score_2'.

simonandras commented on doc/whats_new/v1.1.rst:
```diff
@@ -952,6 +952,16 @@ Changelog
+- |Fix| :class:`mixture.GaussianMixture` now normalizes `weights_` more safely,
+  preventing rounding errors when calling :meth:`GaussianMixture.sample` with
+  `n_components=1`.
+  :pr:`23034` by :user:`Meekail Zain <micky774>`.
```
Added this entry too to resolve merge conflict.

simonandras committed:
Correct doc.

thomasjpfan commented on sklearn/model_selection/_validation.py:
```diff
@@ -758,27 +758,50 @@ def _score(estimator, X_test, y_test, scorer, error_score="raise"):
             scores = scorer(estimator, X_test, y_test)
     except Exception:
-        if error_score == "raise":
+        if isinstance(scorer, _MultimetricScorer):
+            assert error_score == "raise", (
+                "If `_MultimetricScorer` raises exception, the `error_score`"
+                " parameter should be equal to 'raise'."
+            )
```
I think we do not need to check this because we will are going to raise anyways:

```python
except Exception:
    if error_score == "raise":
        raise
    elif not isinstance(scorer, _MultimetricScorer):
        ...
```

thomasjpfan commented on sklearn/model_selection/_validation.py:
```diff
@@ -758,27 +758,50 @@ def _score(estimator, X_test, y_test, scorer, error_score="raise"):
+            (name, str_e) for name, str_e in scores.items() if isinstance(str_e, str)
+        ]
+        if exception_messages:
+            assert error_score != "raise", (
+                "`error_score` == 'raise', but the exception is not raised in"
+                " `_MultimetricScorer`."
```
Having assertions check for logic like this is not too common in the library. I'll be okay with a common stating the `error_score != "raise"` here.

thomasjpfan reviewed (commented):
Thanks for the update. I'm mostly okay with using a string to send the exception information back to the caller. Note other reviewers, may have a different opinion on the matter.

simonandras committed:
Changing changelog based on rewiev.

simonandras committed:
Removing assertions from _validation.py and leaving comments instead.

simonandras commented:
@thomasjpfan I do not understand the failing test. Do you know what is the issue there?

thomasjpfan commented:
Merging with `main` should fix the circleci issue.

simonandras commented:
@thomasjpfan how do you see, do we need any more modification to the changes?

thomasjpfan commented:
Assuming you ran this to set the remote url:

```bash
git remote add upstream git@github.com:scikit-learn/scikit-learn.git
```

you can update this PR to fix the circleci issue:

```bash
git checkout cross_validate__with_multiple_scorer_nan_issue
git fetch upstream
git merge upstream/main
```

simonandras committed:
Merge remote-tracking branch 'upstream/main' into cross_validate__with_multiple_scorer_nan_issue

simonandras commented:
Thank you, it wokred!

thomasjpfan commented on sklearn/metrics/_scorer.py:
```diff
@@ -103,11 +110,18 @@ def __call__(self, estimator, *args, **kwargs):
+                if self._raise_exc:
+                    raise e
+                else:
+                    scores[name] = format_exc()
```
For future reviewers, a string is used to pass the information back to the caller.

thomasjpfan reviewed (approved):
Thank you for the update! I am mostly on board with this. Let's see what other reviewers think.

simonandras commented:
@thomasjpfan could you please mark this as "waiting for review"?

glemaitre committed:
Merge branch 'main' into cross_validate__with_multiple_scorer_nan_issue

glemaitre committed:
DOC move entry changelog

glemaitre committed:
Apply suggestions from code review

glemaitre reviewed (approved):
Since the changes that I requested are only about style, I merged them directly and put my +1.

I will merge when the CI turns green. Thanks @simonandras
